src/file_chooser_dialog.rs

```diff
@@ -5,21 +5,91 @@
 use ffi;
 use glib::translate::*;
 use glib::object::{Downcast, IsA};
+use libc::c_char;
 use std::ptr;
 use FileChooserAction;
 use FileChooserDialog;
+use ResponseType;
 use Widget;
 use Window;
 
 impl FileChooserDialog {
-    pub fn new<T: IsA<Window>>(title: Option<&str>, parent: Option<&T>, action: FileChooserAction)
-            -> FileChooserDialog {
+    // TODO: Keep the other constructor with buttons support as the only constructor (this one was
+    //       left for compatibility) and rename it to `new` for consistency.
+    pub fn new<T: IsA<Window>>(title: Option<&str>, parent: Option<&T>, action: FileChooserAction) -> FileChooserDialog {
         assert_initialized_main_thread!();
         unsafe {
-            Widget::from_glib_none(
-                ffi::gtk_file_chooser_dialog_new(title.to_glib_none().0, parent.to_glib_none().0,
-                    action.to_glib(), ptr::null_mut()))
+            Widget::from_glib_none(ffi::gtk_file_chooser_dialog_new(
+                title.to_glib_none().0,
+                parent.to_glib_none().0,
+                action.to_glib(),
+                ptr::null::<c_char>()
+            ))
                 .downcast_unchecked()
         }
     }
+
+    pub fn with_buttons<T: IsA<Window>>(title: Option<&str>, parent: Option<&T>, action: FileChooserAction, buttons: &[(&str, ResponseType)]) -> FileChooserDialog {
+        assert_initialized_main_thread!();
+        unsafe {
+            Widget::from_glib_none(match buttons.len() {
+                0 => {
+                    ffi::gtk_file_chooser_dialog_new(
+                        title.to_glib_none().0,
+                        parent.to_glib_none().0,
+                        action.to_glib(),
+                        ptr::null::<c_char>()
+                    )
+                },
+                1 => {
+                    let first_button_text: Stash<*const c_char, str> = buttons[0].0.to_glib_none();
+                    ffi::gtk_file_chooser_dialog_new(
+                        title.to_glib_none().0,
+                        parent.to_glib_none().0,
+                        action.to_glib(),
+                        first_button_text.0,
```

Comment:
Why not just `buttons[0].0.to_glib_none().0` here?

Reply:
For consistency with the other branches.

To be honest, I also do not really understand how Stash works